File: src/chapters/interpolation/content/code/newton.cpp
#include <vector>
#include <iostream>
using namespace std;
using Vec = vector<double>;

// Newton divided-difference coefficients a_i = f[x_0,...,x_i].
Vec divided_differences(const Vec& x, const Vec& y) {
    int n = x.size();
    Vec a = y;
    for (int j = 1; j < n; ++j)
        for (int i = n - 1; i >= j; --i)
            a[i] = (a[i] - a[i - 1]) / (x[i] - x[i - j]);
    return a;
}

double newton_eval(const Vec& x, const Vec& a, double t) {
    double p = a.back();
    for (int k = (int)a.size() - 2; k >= 0; --k) p = p * (t - x[k]) + a[k];
    return p;
}

int main() {
    Vec x = {-1, 1, 2, 3}, y = {-3, 1, 3, 29};
    Vec a = divided_differences(x, y);
    cout << "divided differences:";
    for (double v : a) cout << " " << v;
    cout << "\np(0) = " << newton_eval(x, a, 0.0) << "\n";
}

File: src/chapters/interpolation/content/code/newton.rs
// Newton divided-difference coefficients a_i = f[x_0, ..., x_i].
fn divided_differences(x: &[f64], y: &[f64]) -> Vec<f64> {
    let n = x.len();
    let mut a = y.to_vec();
    for j in 1..n {
        for i in (j..n).rev() {
            a[i] = (a[i] - a[i - 1]) / (x[i] - x[i - j]);
        }
    }
    a
}
fn main() {
    let x = [-1.0, 1.0, 2.0, 3.0];
    let y = [-3.0, 1.0, 3.0, 29.0];
    println!("{:?}", divided_differences(&x, &y));   // [-3, 2, 0, 3]
}
